ably/hosts.go

```diff
@@ -0,0 +1,233 @@
+package ably
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ably/ably-go/ably/internal/ablyutil"
+)
+
+// RSC15
+type restHosts struct {
+	primaryFallbackHost string
+	opts                *clientOptions
+	cache               *hostCache
+	visitedHosts        []string
+	sync.Mutex
+}
+
+func newRestHosts(opts *clientOptions) *restHosts {
+	return &restHosts{
+		opts: opts,
+		cache: &hostCache{
+			duration: opts.fallbackRetryTimeout(),
+		},
+	}
+}
+
+func (restHosts *restHosts) getPrimaryHost() string {
+	return restHosts.opts.getPrimaryRestHost()
+}
+
+func (restHosts *restHosts) getFallbackHost() string {
+	restHosts.Lock()
+	defer restHosts.Unlock()
+	hosts, _ := restHosts.opts.getFallbackHosts()
+	shuffledFallbackHosts := ablyutil.Shuffle(hosts)
+	getNonVisitedHost := func() string {
```

Comment:
Why is this a function given it's only called once, right away?

Reply:
It's not about **re-usability**, rather **readability**. Also, that particular snippet exists in the context of returning right fallback host, so makes sense to keep it anonymous, while abstracting internal details